dotnet/src/InternalUtilities/src/Diagnostics/ModelDiagnostics.cs

```diff
@@ -0,0 +1,246 @@
+        T prompt,
+        PromptExecutionSettings? executionSettings,
+        Func<T, string> formatPrompt)
+    {
+        if (!IsModelDiagnosticsEnabled())
+        {
+            return null;
+        }
+
+        string operationName = prompt is ChatHistory ? "chat.completions" : "text.completions";
+        var activity = s_activitySource.StartActivityWithTags(
+            $"{operationName} {modelName}",
+            [
+                new(ModelDiagnosticsTags.Operation, operationName),
+                new(ModelDiagnosticsTags.System, modelProvider),
+                new(ModelDiagnosticsTags.Model, modelName),
+            ]);
+
+        if (endpoint is not null)
+        {
+            activity?.AddTags([
+                new(ModelDiagnosticsTags.Address, endpoint.AbsoluteUri),
+                new(ModelDiagnosticsTags.Port, endpoint.Port),
+            ]);
+        }
+
+        AddOptionalTags(activity, executionSettings);
+
+        if (s_enableSensitiveEvents)
+        {
+            var formattedContent = formatPrompt(prompt);
+            activity?.AttachSensitiveDataAsEvent(
+                ModelDiagnosticsTags.PromptEvent,
+                [
+                    new(ModelDiagnosticsTags.PromptEventPrompt, formattedContent),
+                ]);
+        }
+
+        return activity;
+    }
+
+    /// <summary>
+    /// Set the completion response for a given activity.
+    /// The `formatCompletions` delegate won't be invoked if events are disabled.
+    /// </summary>
+    private static void SetCompletionResponse<T>(
+        Activity activity,
+        T completions,
+        int? promptTokens,
+        int? completionTokens,
+        Func<T, string> formatCompletions) where T : IEnumerable<KernelContent>
+    {
+        if (!IsModelDiagnosticsEnabled())
+        {
+            return;
+        }
+
+        activity.AddTags(
+            [
+                new(ModelDiagnosticsTags.FinishReason, GetFinishReasons(completions)),
+                new(ModelDiagnosticsTags.PromptToken, promptTokens ?? -1),
+                new(ModelDiagnosticsTags.CompletionToken, completionTokens ?? -1),
+                new(ModelDiagnosticsTags.ResponseId, GetResponseId(completions.FirstOrDefault())),
+            ]);
+
+        if (s_enableSensitiveEvents)
+        {
+            activity.AttachSensitiveDataAsEvent(
+                ModelDiagnosticsTags.CompletionEvent,
+                [
+                    new(ModelDiagnosticsTags.CompletionEventCompletion, formatCompletions(completions)),
+                ]);
+        }
+    }
+
+    private static string GetFinishReasons(IEnumerable<KernelContent> completions)
+    {
+        var finishReasons = completions.Select(c =>
+        {
+            if (c.Metadata?.TryGetValue("FinishReason", out var finishReason) == true)
+            {
+                return finishReason;
+            }
+
+            return null;
+        });
+
+        return $"[{string.Join(", ", finishReasons)}]";
```

Comment:
👍 
nit: any reason to do `"[stop, stop]"` instead of `"stop,stop"` ?
`[stop, stop]` is not a valid json, nothing will expect json, in many cases there is just one stop reason. 
Also I'd avoid ` `